src/NServiceBus.RavenDB/Outbox/RavenDbOutboxStorage.cs

```diff
@@ -30,59 +31,79 @@ protected override void Setup(FeatureConfigurationContext context)
             }
 
             protected override Task OnStart(IMessageSession session)
             {
                 timeToKeepDeduplicationData = settings.GetOrDefault<TimeSpan?>("Outbox.TimeToKeepDeduplicationData") ?? TimeSpan.FromDays(7);
 
-                var frequencyToRunDeduplicationDataCleanup = settings.GetOrDefault<TimeSpan?>("Outbox.FrequencyToRunDeduplicationDataCleanup") ?? TimeSpan.FromMinutes(1);
+                frequencyToRunDeduplicationDataCleanup = settings.GetOrDefault<TimeSpan?>("Outbox.FrequencyToRunDeduplicationDataCleanup") ?? TimeSpan.FromMinutes(1);
+
+                cancellationTokenSource = new CancellationTokenSource();
+                cancellationToken = cancellationTokenSource.Token;
 
-                cleanupTimer = new Timer(PerformCleanup, null, TimeSpan.FromMinutes(1), frequencyToRunDeduplicationDataCleanup);
+                cleanupTask = Task.Run(() => PerformCleanup(), CancellationToken.None);
 
                 return TaskEx.CompletedTask;
             }
 
-            protected override Task OnStop(IMessageSession session)
+            protected override async Task OnStop(IMessageSession session)
             {
-                using(var waitHandle = new ManualResetEvent(false))
+                cancellationTokenSource.Cancel();
+
+                // ReSharper disable once MethodSupportsCancellation
+                var timeoutTask = Task.Delay(TimeSpan.FromSeconds(30));
+                var finishedTask = await Task.WhenAny(cleanupTask, timeoutTask).ConfigureAwait(false);
+
+                if (finishedTask == timeoutTask)
                 {
-                    cleanupTimer.Dispose(waitHandle);
-                    waitHandle.WaitOne();
-                    cleanupTimer = null;
+                    logger.Error("RavenOutboxCleaner failed to stop within the time allowed (30s).");
                 }
-
-                return TaskEx.CompletedTask;
             }
 
-            void PerformCleanup(object state)
+            async Task PerformCleanup()
             {
-                cleaner.RemoveEntriesOlderThan(DateTime.UtcNow - timeToKeepDeduplicationData)
-                    .GetAwaiter()
-                    .GetResult();
+                while (!cancellationToken.IsCancellationRequested)
+                {
+                    var nextClean = DateTime.UtcNow.Add(frequencyToRunDeduplicationDataCleanup);
+
+                    try
+                    {
+                        var olderThan = DateTime.UtcNow - timeToKeepDeduplicationData;
+                        await cleaner.RemoveEntriesOlderThan(olderThan).ConfigureAwait(false);
+                    }
+                    catch (OperationCanceledException)
+                    {
+                        // Graceful shutdown
+                    }
+                    catch (Exception ex)
+                    {
+                        logger.Warn("Unable to remove expired Outbox records from Raven database.", ex);
```

Comment:
Perhaps, or at the very least log ERROR instead of WARN?

At some point this little problem will escalate into a really big problem.